tokio-timer/src/debounce.rs

```diff
@@ -0,0 +1,397 @@
+    /// pulling from.
+    pub fn get_ref(&self) -> &T {
+        &self.stream
+    }
+
+    /// Acquires a mutable reference to the underlying stream that this combinator
+    /// is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the stream
+    /// which may otherwise confuse this combinator.
+    pub fn get_mut(&mut self) -> &mut T {
+        &mut self.stream
+    }
+
+    /// Consumes this combinator, returning the underlying stream.
+    ///
+    /// Note that this may discard intermediate state of this combinator, so care
+    /// should be taken to avoid losing resources when this is called.
+    pub fn into_inner(self) -> T {
+        self.stream
+    }
+
+    /// Computes the instant at which the next debounce delay elapses.
+    fn delay_time(&mut self) -> Instant {
+        let next = clock::now() + self.duration;
+
+        if let Some(to) = self.max_wait_to {
+            cmp::min(next, to)
+        } else {
+            next
+        }
+    }
+
+    /// Polls the underlying delay future.
+    fn poll_delay(d: &mut Delay) -> Poll<(), <Self as Stream>::Error> {
+        d.poll().map_err(DebounceError::from_timer_error)
+    }
+
+    /// Polls the underlying stream.
+    fn poll_stream(
+        &mut self,
+    ) -> Poll<Option<<Self as Stream>::Item>, <Self as Stream>::Error> {
+        self.stream.poll().map_err(DebounceError::from_stream_error)
+    }
+
+    /// Starts a new delay using the current duration and maximum waiting time.
+    fn start_delay(&mut self) {
+        self.max_wait_to = self.max_wait.map(|dur| clock::now() + dur);
+        self.delay = Some(Delay::new(self.delay_time()));
+    }
+}
+
+impl<T: Stream> Stream for Debounce<T> {
+    type Item = T::Item;
+    type Error = DebounceError<T::Error>;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.delay.take() {
+            Some(mut d) => match Self::poll_delay(&mut d)? {
+                // Delay has woken us up and is over, if we're trailing edge this
+                // means we need to return the last item.
+                Async::Ready(_) => {
+                    if self.edge.is_trailing() {
+                        if let Some(item) = self.last_item.take() {
+                            return Ok(Async::Ready(Some(item)));
+                        }
+                    }
+
+                    return Ok(Async::NotReady);
+                },
+
+                // The stream has woken us up, but we have a delay.
+                Async::NotReady => match self.poll_stream()? {
+                    // We have gotten an item, but we're currently blocked on
+                    // the delay. Save it for later and reset the timer.
+                    Async::Ready(Some(item)) => {
+                        d.reset(self.delay_time());
+
+                        self.delay = Some(d);
+                        self.last_item = Some(item);
```

Comment:
I guess the problem is that we aren't being woken up "correctly" here. Since there is no polling between the first and the second item, there is no processing transitioning us from storing the first to storing the second.

However, if you add a 

```rust
assert_not_ready!(debounced);
```

between pushing the first and the second item into the stream, everything works out as expected, because the debouncer has the chance to process the new item.

I wonder how we can solve this, and even if we _need to_ solve this. To me, this seems more like an architectural issue or quirk with pull-streams rather than with this implementation. In a real-world-application one would have the debouncer running on an executor, which should wake it up two times for both items being sent through, shouldn't it?

I mean, we could make the polling recursive in some cases, i. e. poll ourselves again after we have started the delay and returned `NotReady`. That would solve this specific case (with a finite amount of elements buffered), but obviously lead to issues with streams that are an infinite source of items like `stream::repeat()`.

Any other ideas?

Reply:
> If we don't care about that, then we could leave it as is... although we'd end up with more of a token bucket sort of thing, I think. Since we only grab one item max before triggering the debounce delay, if there was a burst of N items, we'd actually end up debouncing N times, which depending on your duration, could be a really drawn out thing.

Doesn't this depend on how often we're woken up? Wouldn't any regular implementation wake us up N times as well?

> At least for the standard streams like an mpsc::Receiver, it follows that contract. That's why polling until NotReady is generally important.

Am I reading this correctly that sending into an `mpsc::Receiver` N times does not wake the stream up N times?

> I think it depends on what behavior we want to expose here. If our idea of debouncing is tied to wall time, for example, if I send the first item at t=0, and we want to debounce until t=10, then items sent between 0-10 should wake me up and affect the debounce... so we always need to poll the underlying stream until NotReady to have it continue to wake us up when it has another item to give after being empty.

I think we should go for wall clock time.

> As far as something like stream::repeat, yeah, no great answer there. Although, even with push streams, we would just endlessly be pushing as fast as possible, so one way or another, stream::repeat is a weird degenerate case that I'm not sure we could ever handle. max_wait could help us, but we'd be waking up or spinning endlessly whether it's push vs pull.

We could document that the underlying stream needs to be `NotReady` at least some times, that'd be okay I think.

@carllerche what's your opinion on this?

Reply:
> Doesn't this depend on how often we're woken up? Wouldn't any regular implementation wake us up N times as well?

It does, yes.  As the code stands, you'll only be woken up by a delay reaching its timeout, or if `poll_stream` is called and `NotReady` is returned, signaling that the underlying will notify us against when there is an item to emit.

> Am I reading this correctly that sending into an `mpsc::Receiver` N times does not wake the stream up N times?

Yes.  You can roughly see it in action here: https://docs.rs/futures/0.1.25/src/futures/sync/mpsc/mod.rs.html#931

If there's an available message, it's returned.  If there's no message available, the task (the one receiving) is parked and notified when an item is sent in.  So, if you don't drain the receiver until `NotReady` is reached, the receiver task is never scheduled to get a wakeup if an item is sent.
 
> I think we should go for wall clock time, i. e. change the impl to poll until we see `NotReady`.

Makes sense to me.

> We could document that the underlying stream needs to be `NotReady` at least some times, that'd be okay I think.

That sounds ideal to me.

Reply:
Done.